gst/nnstreamer/elements/gsttensor_trainer.c

```diff
@@ -627,126 +620,317 @@ gst_tensor_trainer_epochs_is_complete (GstTensorTrainer * trainer)
-  void *ptr;
-
-  trainer = GST_TENSOR_TRAINER (parent);
-
-  in_flexible = gst_tensor_pad_caps_is_flexible (sinkpad);
-  num_tensors = gst_tensor_buffer_get_count (inbuf);
+  if (trainer->is_training_complete == TRUE) {
+    /** app need to send gst_element_send_event(tensor_trainer, gst_event_new_eos())
+        after training_complete or set eos to datareposrc */
+    GST_WARNING_OBJECT (trainer,
+        "Training is completed, buffer is dropped, please change state of pipeline");
+    return TRUE;
+  }
+  return FALSE;
+}
 
+/**
+ * @brief  Check chain conditions. If all conditions are met, proceed to next step.
+ */
+static gboolean
+gst_tensor_trainer_check_chain_conditions (GstTensorTrainer * trainer,
+    guint num_tensors)
+{
   if (!trainer->fw_created) {
     if (!gst_tensor_trainer_check_invalid_param (trainer))
-      return GST_FLOW_ERROR;
+      return FALSE;;
     if (!gst_tensor_trainer_create_model (trainer))
-      return GST_FLOW_ERROR;
+      return FALSE;
   }
 
   if (num_tensors >= NNS_TENSOR_SIZE_LIMIT)
-    return GST_FLOW_ERROR;
+    return FALSE;
 
-  if (trainer->is_training_complete == TRUE) {
-    /** app need to send gst_element_send_event(tensor_trainer, gst_event_new_eos())
-        after training_complete or set eos to datareposrc */
-    GST_WARNING_OBJECT (trainer,
-        "Training is completed, buffer is dropped, please change state of pipeline");
-    return GST_FLOW_OK;
+  return TRUE;
+}
+
+/**
+ * @brief Get the size of tensor header.
+ */
+static gsize
+gst_tensor_trainer_get_header_size (GstTensorTrainer * trainer,
+    gboolean in_flexible, GstTensorInfo * info, GstTensorMetaInfo * meta,
+    void *data, int i)
+{
+  gsize header_size = 0;
+
+  if (in_flexible) {
+    if (!gst_tensor_meta_info_parse_header (meta, data)) {
+      GST_ERROR_OBJECT (trainer, "Invalid Flexible tensors");
+      return 0;
+    }
+
+    header_size = gst_tensor_meta_info_get_header_size (meta);
+    info = gst_tensors_info_get_nth_info (&trainer->prop.input_meta, i);
+    if (info == NULL)
+      return 0;
+    gst_tensor_meta_info_convert (meta, info);
+    GST_INFO ("flexible header size:%zd", header_size);
+  } else {
+    GST_INFO ("not flexible header, size:%zd", header_size);
   }
 
+  return header_size;
+}
+
+
+/**
+ * @brief Create input tensors from the buffer.
+ */
+static gboolean
+gst_tensor_trainer_create_input_tensors (GstTensorTrainer * trainer,
+    GstBuffer * inbuf, GstMemory ** in_mem, GstMapInfo * in_info,
+    gboolean in_flexible)
+{
+  guint i;
+  GstTensorInfo *info = NULL;
+  GstTensorMetaInfo in_meta[NNS_TENSOR_SIZE_LIMIT];
+  guint num_tensors = 0;
+  gsize header_size;
+
+  num_tensors = gst_tensor_buffer_get_count (inbuf);
+
   for (i = 0; i < num_tensors; i++) {
     in_mem[i] = gst_tensor_buffer_get_nth_memory (inbuf, i);
-    if (!gst_memory_map (in_mem[i], &in_info[i], GST_MAP_READ)) {
+    if (!gst_memory_map (in_mem[i], (in_info + i), GST_MAP_READ)) {
```

Comment:
Is there any reason to change the pointer operation like this?

Reply:
I will change ```((in_info + i)``` to ```&in_info[i]```